stan/math/fwd/core/operator_division.hpp

```diff
@@ -41,11 +44,59 @@ inline fvar<T> operator/(const fvar<T>& x1, double x2) {
  * @param x2 second argument
  * @return first argument divided by second argument
  */
-template <typename T>
-inline fvar<T> operator/(double x1, const fvar<T>& x2) {
-  // TODO(carpenter): store x1 / x2.val_ and reuse
+template <typename T, typename U, require_arithmetic_t<U>...>
+inline fvar<T> operator/(U x1, const fvar<T>& x2) {
   return fvar<T>(x1 / x2.val_, -x1 * x2.d_ / (x2.val_ * x2.val_));
 }
+
+template <typename T>
+inline std::complex<fvar<T>> operator/(const std::complex<fvar<T>>& x1,
+                                       const std::complex<fvar<T>>& x2) {
+  return internal::complex_divide(x1, x2);
+}
```

Comment:
Could these all be replaced by

```cpp
template <typename... Types>
using require_any_complex_t = require_any_t<is_complex<Types>...>;

template <typename T1, typename T2, typename = require_any_complex_t<T1, T2>>
inline auto operator/(const  T1& x, const T2& y) {
    return internal::complex_divide(x, y);
}

Reply:
There are built-ins in the standard library for `(complex<T>, complex<T>)` and `(T, complex<T>)` and `(complex<T>, T)` that cause all kinds of ambiguity problems if we don't specify our types more specifically.  The built-ins don't work for g++ because they don't support simple class definitions of `std::complex` (they assume member variables that don't exist, for example).  Also, the pattern we want to match is `any_complex` *and* `any_autodiff`---you need one autodiff type and one complex type in any combination.

Reply:
If it's both

1. Any argument being complex
2. Any argument being an autodiff type

```cpp
template <typename... Types>
using require_any_complex_t = require_any_t<is_complex<Types>...>;

template <typename T1, typename T2, 
 typename = require_any_complex_t<T1, T2>,
 typename = require_any_autodiff<T1, T2, value_type_t<T1>, value_type_t<T2>>
inline auto operator/(const  T1& x, const T2& y) {
    return internal::complex_divide(x, y);
}
```

So that would only go off if either T1 or T2 is a complex type and if T1 or T2 or it's value_types are autodiff types.

Unless that can be used in a bunch of places and deletes lotsa code then it's fine as is

Reply:
The problem isn't restricting matches, but ambiguity with the `std:: version` when it does match.  The `std::complex` versions get brough in by ADL because `complex` is a class.

The specific problem here are these system definitions:

```
namespace std {
template <typename T>
complex<T> operator/(complex<T> a, complex<T> b);

template <typename T>
complex<T> operator/(T a, complex<T> b);

template <typename T>
complex<T> operator/(complex<T> a, T b);
}
```

These get brought in by ADL.  They only match cases where the real and complex have the same type.  And this is all that the standard library gives you, which is one of the reasons we have to go beyond it to allow mixing of `double` and autodiff types at all levels.  If we try to cover those cases generically, we get conflicts with the standard library.  At least any way I've tried it

And before trying to types traits away the `T1 == T2` case with `std::is_same`, let me emphasize again that we can't have the standard library function version matching because their implementations vary between clang++ and g++ and aren't compatible.  So we need to make sure our version of `operator/` gets called.  That's why this is all getting defined in the first place---relying on system specializations of `complex<T>` is undefined behavior!

So what we need is specific enough signatures defined in `stan::math` to provide beter matches than the system built-ins.  Now the system versions vary between clang++ and g++ (that's what they mean by undefined behavior!), as I found when originally going down the route you're trying to go down with fewer specializations.

If you want to simplify, I'd suggest checking out and running through all the tests online (even the Linux MPI tests bring in different library functions than the usual clang++ and g++, which was the problem with the last test).

You'll also see  what look like useless header includes.  Those are there for the header tests because the function calls other functions internally that need `rev` definitions in scope.  Because everything's templated for forward-mode, the header tests aren't so strict because nothing's getting instantiated enough to test.

Also, my final version only uses overloads in the `stan::math` namespace.  My original solution had a lot of specializations in the `std::` namespace.  That's no longer necessary with the richer set of overloads.  In some cases, the number of overloads can be cut down because both g++ and clang++ are missing specializations or template versions.  But I'd rather not rely on their undefined behavior for template matching.